### PR TITLE
OPENEUROPA-1503: Add content lock module.

### PR DESCRIPTION
**Pull request by imanoleguskiza**
### Description

Provide a module that prevents content from being edited by two editors at the same time.

### Change log

- Added: oe_editorial_content_lock module

### CODE REVIEW

File: tests/features/content_lock.feature [hunk 1]
```diff
@@ -0,0 +1,15 @@
+@api @javascript
+Feature: Content lock
+  As a content editor
```
[the remark this answers is not in the document]
Comment (imanoleguskiza) [reply tests/features/content_lock.feature]: I'm not sure I agree with this change since its not only being notified but actively prevent other users from editing it

File: tests/Behat/FeatureContext.php [hunk 2]
```diff
@@ -15,6 +15,13 @@
  */
 class FeatureContext extends RawDrupalContext {
 
+  /**
+   * Keep track of modules so they can be cleaned up.
+   *
+   * @var array
+   */
+  protected $modules = [];
```
Comment (voidtek) [inline tests/Behat/FeatureContext.php]: We are not using this variable.

File: tests/Behat/FeatureContext.php [hunk 3]
```diff
@@ -93,4 +100,30 @@ protected function assertElementType(NodeElement $element, string $type): void {
...
+   */
+  public function disableContentLock() {
+    $modules = [
+      'oe_editorial_content_lock',
+      'content_lock',
+    ];
+    foreach ($modules as $module) {
+      \Drupal::service('module_installer')->uninstall([$module]);
+    }
+    $this->modules = [];
```
Comment (voidtek) [inline tests/Behat/FeatureContext.php]: This line is not useful.

File: modules/oe_editorial_content_lock/oe_editorial_content_lock.install [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+declare(strict_types = 1);
+
+/**
+ * Implements hook_install().
+ */
+function oe_editorial_content_lock_install() {
+  // Set default settings for Content Lock.
+  $locked_contents = [
+    'file' => [
+      '*' => '*',
```
Comment (voidtek) [inline modules/oe_editorial_content_lock/oe_editorial_content_lock.install]: This kind of contents is not tested on the behat test.

File: tests/Behat/FeatureContext.php [hunk 5]
```diff
@@ -93,4 +100,30 @@ protected function assertElementType(NodeElement $element, string $type): void {
     }
   }
 
+  /**
+   * Install the content_lock component.
+   *
+   * @BeforeScenario @content_lock
+   */
+  public function enableContentLock() {
+    // Revert config that was changed.
```
Comment (voidtek) [inline tests/Behat/FeatureContext.php]: This comment is not correct.

File: tests/features/content_lock.feature (hunk 1 above)
[the remark this answers is not in the document]
Comment (brummbar) [reply tests/features/content_lock.feature]: Maybe
```
as a content editor
I should be prevented from concurrently editing a node
so that no work is lost
``` 

like in the readme?

**Comment by voidtek** [comment]
The behat functions associated with locked_content can be on a different files (e.g. tests/Behat/LockedContentContext.php)

File: modules/oe_editorial_content_lock/oe_editorial_content_lock.install (hunk 4 above)
Comment (imanoleguskiza) [reply modules/oe_editorial_content_lock/oe_editorial_content_lock.install]: I updated your other comments but I'm not sure how to tackle this. I didn't find any means of editing a file in core D8 so I'm not sure if we should even include it. What do you think?

File: tests/Behat/ContentLockContext.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+  /**
+   * Uninstall the content_lock component.
+   *
+   * @AfterScenario @content_lock
+   */
+  public function disableContentLock() {
+    $modules = [
+      'oe_editorial_content_lock',
+      'content_lock',
```
Comment (sergepavle) [inline tests/Behat/ContentLockContext.php]: why are we including in dependencies module content_lock but after finishing tests uninstalling this module?

Comment (imanoleguskiza) [reply tests/Behat/ContentLockContext.php]: When we enable oe_content_lock on the enableContentLock function this enables content_lock as its a dependency. When we cleanup once the scenario is finished, both modules need to be disabled.

File: tests/Behat/ContentLockContext.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    ];
+    // Remove the Content Lock feature and its dependencies.
+    foreach ($modules as $module) {
+      \Drupal::service('module_installer')->uninstall([$module]);
+    }
+  }
+
+}
```
Comment (ademarco) [inline tests/Behat/ContentLockContext.php]: I would remove this entirely, we enable it via the task runner, we should use this method only when we are testing optional sub-modules.

File: modules/oe_editorial_content_lock/oe_editorial_content_lock.install [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  // Set default settings for Content Lock.
+  $locked_contents = [
+    'node' => [
+      '*' => '*',
+    ],
+  ];
+  \Drupal::configFactory()
+    ->getEditable('content_lock.settings')
+    ->set('types', $locked_contents)
+    ->save();
```
Comment (ademarco) [inline modules/oe_editorial_content_lock/oe_editorial_content_lock.install]: Remember to always add this at the beginning of hook_install() when we update existing configuration within it:
```
  if (\Drupal::isConfigSyncing() === TRUE) {
    // If config is syncing, we do nothing here.
    return;
  }
```
The reason being that , when configuration is syncing (i.e. when running `drush cim`) the configuration object that we are trying to modify (in this case `content_lock.settings`) might not exist yet.

File: tests/features/content_lock.feature [hunk 9]
```diff
@@ -0,0 +1,16 @@
+@api @javascript
+Feature: Content lock
+  as a content editor
+  I should be prevented from concurrently editing a node
+  so that no work is lost
+
+  @content_lock
+  Scenario: Content gets lock and lock can be broken
+    Given I am logged in as a user with the "Author" role
+    And I am viewing my "oe_workflow_demo" with the title "Demo site"
```
Comment (ademarco) [inline tests/features/content_lock.feature]: This step should be more business readable, we could also create a custom one if needed.

File: tests/Behat/ContentLockContext.php (hunk 7 above)
Comment (imanoleguskiza) [reply tests/Behat/ContentLockContext.php]: But the content_lock module doesn't get enabled on the task runner as of now, it is an optional module. If we enable it by default in the task runner I think it will start screwing with all the other tests

Comment (ademarco) [reply tests/Behat/ContentLockContext.php]: You are right! Sorry about that, somehow I missed it! :+1:

File: tests/features/content_lock.feature (hunk 9 above)
Comment (ademarco) [reply tests/features/content_lock.feature]: My problem is mostly with `oe_workflow_demo`, as in: no business person will know what we refer to. Maybe something like this?

```gherkin
And I am viewing a content page with title "Demo site"
```

**Comment by nagyad** [comment]
Good job. Regarding the behat test, please extend it with an appropriate user story where 2 simultaneous editing are happening and check that the Save is disabled for the locked one.
Thank you!

File: tests/Behat/ContentLockContext.php [hunk 10]
```diff
@@ -42,6 +38,58 @@ public function createMyNode($type_name, $title): void {
...
+   *
+   * @param string $type_name
+   *   The content type of the node.
+   * @param string $title
+   *   The title of the node.
+   *
+   * @throws \Exception
+   *   Thrown when node is not found.
+   *
+   * @Then I visit the :type_name content titled :title
```
Comment (nagyad) [inline tests/Behat/ContentLockContext.php]: How about this? Then I visit the :type_name content with title: :title ?

File: tests/Behat/ContentLockContext.php [hunk 11]
```diff
@@ -68,4 +116,42 @@ public function disableContentLock(): void {
...
   }
 
+  /**
+   * Remove any created nodes.
+   *
+   * @AfterScenario
+   */
+  public function cleanNodes(): void {
+    if (\Drupal::moduleHandler()->moduleExists('content_lock')) {
+      \Drupal::database()->delete('content_lock')
```
Comment (nagyad) [inline tests/Behat/ContentLockContext.php]: You are not cleaning nodes here, right? Then we should not indicate we do.

File: tests/Behat/ContentLockContext.php (hunk 10 above)
Comment (imanoleguskiza) [reply tests/Behat/ContentLockContext.php]: I wanted to keep it consistent with the other steps provided by the context, but I don't oppose this :)

File: tests/Behat/ContentLockContext.php (hunk 11 above)
Comment (imanoleguskiza) [reply tests/Behat/ContentLockContext.php]: I'm overriding the cleanNodes provided by the parent class though :(

File: tests/Behat/ContentLockContext.php [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+namespace Drupal\Tests\oe_editorial\Behat;
+
+use Drupal\DrupalExtension\Context\DrupalContext;
+use PHPUnit\Framework\Assert;
+
+/**
+ * Defines step definitions for testing the Content Lock feature.
+ */
+class ContentLockContext extends DrupalContext {
```
Comment (nagyad) [inline tests/Behat/ContentLockContext.php]: Why did you change this from RawDrupalContext?

**Comment by pfrenssen** [comment]
In this PR content locking is being introduced, and this is causing a problem during the automated cleanup of content created in Behat scenarios. In order to keep the test environment reasonably clean, Behat keeps track of the content that is created in a scenario and will remove it after the scenario is completed. This happens in the "AfterScenario" hook which is an event with that is tagged with `@AfterScenario` in an annotation.

The problem that occurs is that if content is locked then it is possible the deletion of the content is being denied during the automated cleanup. A possible solution would be to implement our own `AfterScenario` that will unlock content before it is deleted, but at this time it does not seem possible to guarantee that our hook will run before the one provided by the Drupal Extension. Behat doesn't appear to allow us to set a priority for our hook.

The right solution would be to implement support for hook priorities in Behat, but this would be a non-trivial task. The effort required for this is in my opinion not justifiable in scope of this ticket.

I would propose to solve this right now with a simple workaround. Let's define a step `@Then I unlock all locked content` which can be put as the last line of the affected scenarios.

If we have another situation in the future which requires hook priorities in Behat we should create a ticket and do the effort required to implement this so that there is a structural solution which can be used across all our projects. For the moment though this looks like an edge case.

File: .drone.yml [hunk 13]
```diff
@@ -11,6 +11,11 @@ services:
     image: percona/percona-server:5.6
     environment:
       - MYSQL_ALLOW_EMPTY_PASSWORD=yes
+  selenium:
+    image: selenium/standalone-chrome:3.11
```
Comment (upchuk) [inline .drone.yml]: Why are we locking selenium at 3.11? Could/should be 3.

File: modules/oe_editorial_content_lock/README.md [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+By installing this module, a user that starts to edit content on the site will lock it preventing other users from
+editing it at the same time.
+
+This lock can only be broken by the user that created it or by a user with the "Break content lock" permission, and it
+can be done by either saving the content or by clicking the available links on the content editing page.
+
+## Setup content lock for content
+
+The module ships with default configuration that will be set when enabled. This will apply the content locking to
+all content types and file types.
```
Comment (upchuk) [inline modules/oe_editorial_content_lock/README.md]: It does not apply to file types.

File: modules/oe_editorial_content_lock/README.md [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+
+The default configuration can be changed at any time by navigating to the module's configuration page
```
Comment (upchuk) [inline modules/oe_editorial_content_lock/README.md]: This second paragraph does not belong here. Our readme's should not contain info on how the UI can be used to configure stuff.

File: modules/oe_editorial_content_lock/oe_editorial_content_lock.install [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+ * Implements hook_install().
+ */
+function oe_editorial_content_lock_install(): void {
+  if (\Drupal::isConfigSyncing() === TRUE) {
+    // If config is syncing, we do nothing here.
+    return;
+  }
+
+  // Set default settings for Content Lock.
+  $locked_contents = [
```
Comment (upchuk) [inline modules/oe_editorial_content_lock/oe_editorial_content_lock.install]: `$locked_content`

File: tests/Behat/ContentLockContext.php [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+
+  /**
+   * Creates content authored by the current user.
+   *
+   * @throws \Exception
+   *   Throws an exception when the content type or the node are not found.
+   *
+   * @Given I am viewing my :type_name content titled :title
+   */
+  public function createMyNode($type_name, $title): void {
```
Comment (upchuk) [inline tests/Behat/ContentLockContext.php]: `createAndViewNodeWithTitle`

Also, strict types

File: tests/Behat/ContentLockContext.php [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+   *   The content type of the node.
+   * @param string $title
+   *   The title of the node.
+   *
+   * @throws \Exception
+   *   Thrown when node is not found.
+   *
+   * @Then I visit the :type_name content titled :title
+   */
+  public function visitContentWithTitle($type_name, $title): void {
```
Comment (upchuk) [inline tests/Behat/ContentLockContext.php]: strict types

File: tests/Behat/ContentLockContext.php [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+
+  /**
+   * Asserts the button with specified id|name|title|alt|value is disabled.
+   *
+   * @throws \Exception
+   *   Thrown when the button is not found.
+   *
+   * @Then the :button button is disabled
+   */
+  public function buttonIsDisabled($button) {
```
Comment (upchuk) [inline tests/Behat/ContentLockContext.php]: strict type and return type

File: tests/Behat/ContentLockContext.php [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+   * @AfterScenario @content_lock
+   */
+  public function disableContentLock(): void {
+    $modules = [
+      'oe_editorial_content_lock',
+      'content_lock',
+    ];
+    // Remove the Content Lock feature and its dependencies.
+    foreach ($modules as $module) {
+      \Drupal::service('module_installer')->uninstall([$module]);
```
Comment (upchuk) [inline tests/Behat/ContentLockContext.php]: What about the config changes it imports and we alter via our `hook_install`?

File: tests/features/content_lock.feature [hunk 21]
```diff
@@ -0,0 +1,23 @@
+@api @javascript
+Feature: Content lock
+  as a content editor
```
Comment (upchuk) [inline tests/features/content_lock.feature]:
```
As a content editor
I should not be able to edit a node
If that node is being edited by someone else
```

File: .drone.yml (hunk 13 above)
Comment (imanoleguskiza) [reply .drone.yml]: https://github.com/search?q=org%3Aopeneuropa+image%3A+selenium%2Fstandalone-chrome-debug&type=Code
We don't seem to have a consensus at the moment :)

File: tests/Behat/ContentLockContext.php (hunk 20 above)
Comment (imanoleguskiza) [reply tests/Behat/ContentLockContext.php]: We are only changing config belonging to the content_lock module. Wouldn't this get deleted when we uninstall content_lock?

File: .drone.yml (hunk 13 above)
Comment (upchuk) [reply .drone.yml]: In absence of a consensus, the latest version is typically preferred no?

File: tests/Behat/ContentLockContext.php (hunk 20 above)
Comment (upchuk) [reply tests/Behat/ContentLockContext.php]: Yes, but that config doesn't get deleted when the content_lock module gets uninstalled because it has no dependencies. Unlike the optional view config that would because it has a dependency on it.
